### ISSUE TITLE
thread 'main' panicked at 'called `Result::unwrap()` ... on :latest 

### ISSUE THREAD

**Issue by scottsweb** [joseluisq/static-web-server#246]
**Describe the bug**

My container auto-updated today and then proceeded to crash. It seems any container in the `2.2x` series is failing for me on armv6 with the following message:

```
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:418:72
```

**To Reproduce**

Docker pull the latest container and start it with the following compose file:

```
version: "3.3"

services:
  web:
    image: joseluisq/static-web-server:2.21
    container_name: web
    env_file: ./.settings.env
    ports:
      - 8080:80
    volumes:
      - ./web:/public:ro
    restart: unless-stopped
    command: -g info
```

settings.env contains the following values:

```
SERVER_PORT=80
SERVER_ROOT=/public
SERVER_ERROR_PAGE_404=/public/404.html
```

**Expected behavior**

The container not crash. 

**Logs/Screenshots**

```
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:418:72
```

**Environment and Specs**
 - static-web-server: 2.21
 - OS: Raspberry Pi OS armv6
 - Docker: 20.10.17
 - Client: N/A

**Comment by joseluisq** [comment]
I gave it a quick try in an ARM64 using `linux/arm/v6` and works for me fine using the latest SWS version.

```sh
docker run -it --rm -p 8080:80 --platform linux/arm/v6 joseluisq/static-web-server:2.20.1-alpine uname -a
# Unable to find image 'joseluisq/static-web-server:2.20.1-alpine' locally
# 2.20.1-alpine: Pulling from joseluisq/static-web-server
# 33ec62e98cee: Pull complete
# f19c8f4d8a6a: Pull complete
# 669dd6581d2e: Pull complete
# 6e5f682ad5e4: Pull complete
# a37d07888f06: Pull complete
# Digest: sha256:167a4cfeb614ac2f0b3dddc4bd4e317cff8b837a58558ba318552a4624a5b045
# Status: Downloaded newer image for joseluisq/static-web-server:2.20.1-alpine
# Linux c80373ec6127 5.15.49-linuxkit-pr #1 SMP PREEMPT Thu May 25 07:27:39 UTC 2023 armv7l Linux
```

```sh
docker run -it --rm -p 8080:80 --platform linux/arm/v6 joseluisq/static-web-server:2.20.1-alpine -g info
# 2023-07-21T09:53:14.479824Z  INFO static_web_server::info: static-web-server 2.20.1
# 2023-07-21T09:53:14.481297Z  INFO static_web_server::info: log level: info
# 2023-07-21T09:53:14.481995Z  INFO static_web_server::info: server bound to tcp socket [::]:80
# 2023-07-21T09:53:14.482087Z  INFO static_web_server::info: runtime worker threads: 4
# 2023-07-21T09:53:14.482177Z  INFO static_web_server::info: runtime max blocking threads: 512
# 2023-07-21T09:53:14.483030Z  INFO static_web_server::info: fallback page: enabled=false, value=""
# 2023-07-21T09:53:14.483125Z  INFO static_web_server::info: security headers: enabled=false
# 2023-07-21T09:53:14.483203Z  INFO static_web_server::info: auto compression: enabled=true
# 2023-07-21T09:53:14.483280Z  INFO static_web_server::info: compression static: enabled=false
# 2023-07-21T09:53:14.483352Z  INFO static_web_server::info: directory listing: enabled=false
# 2023-07-21T09:53:14.483422Z  INFO static_web_server::info: directory listing order code: 6
# 2023-07-21T09:53:14.483490Z  INFO static_web_server::info: directory listing format: Html
# 2023-07-21T09:53:14.483571Z  INFO static_web_server::info: cache control headers: enabled=true
# 2023-07-21T09:53:14.483848Z  INFO static_web_server::info: basic authentication: enabled=false
# 2023-07-21T09:53:14.483927Z  INFO static_web_server::info: log remote address: enabled=false
# 2023-07-21T09:53:14.484001Z  INFO static_web_server::info: redirect trailing slash: enabled=true
# 2023-07-21T09:53:14.484073Z  INFO static_web_server::info: ignore hidden files: enabled=false
# 2023-07-21T09:53:14.484151Z  INFO static_web_server::info: grace period before graceful shutdown: 0s
# 2023-07-21T09:53:14.484246Z  INFO static_web_server::info: health endpoint: enabled=false
# 2023-07-21T09:53:14.489352Z  INFO Server::start_server{addr_str="[::]:80" threads=4}: static_web_server::server: close time.busy=0.00ns time.idle=299µs
# 2023-07-21T09:53:14.490685Z  INFO static_web_server::info: http1 server is listening on http://[::]:80
# 2023-07-21T09:53:14.490776Z  INFO static_web_server::info: press ctrl+c to shut down the server
# 2023-07-21T09:53:31.647948Z  INFO static_web_server::handler: incoming request: method=HEAD uri=/
```

```sh
curl -I http://localhost:8080/
# HTTP/1.1 200 OK
# content-length: 536
# content-type: text/html
# accept-ranges: bytes
# last-modified: Thu, 20 Jul 2023 21:57:48 GMT
# vary: accept-encoding
# cache-control: public, max-age=86400
# date: Fri, 21 Jul 2023 09:53:31 GMT
```

--

I'm suspecting that this is most likely a Docker-related thing but not sure. Because looks like the container isn't allowed to get the current time.

Can you please try the Alpine or Debian variant and share how it goes after?

If that is a Docker thing then it could be related to https://github.com/moby/moby/issues/40734

**Comment by joseluisq** [comment]
Also I see that some people suggest some workaround like running the container in privileged mode or using `--security-opt seccomp=unconfined`.

**Comment by scottsweb** [comment]
I just tested `2-alpine` and that fails with the same log:

```
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }', library/std/src/sys/unix/time.rs:418:72
```

I will test the security opt change next.

**Comment by scottsweb** [comment]
Tested adding `security-opt` to my compose like so:

```
version: "3.3"

services:
  web:
    image: joseluisq/static-web-server:latest
    container_name: web
    env_file: ./.settings.env
    security_opt:
      - seccomp:unconfined
    ports:
      - 8080:80
    volumes:
      - ./web:/public:ro
    restart: unless-stopped
    command: -g info
 ```
 
 And the container starts as expected. So I think this can be closed as this looks to be Docker related.

**Comment by scottsweb** [comment]
Thanks for your help and speedy reply!

**Comment by joseluisq** [comment]
I just quote below what I think should be worth it to note.

> Running with --privileged is not a solution; the reason it works with --privileged is that it disables all (or at least most) security measures that containers provide (among which, disabling seccomp).
>
> I think the problem occurred when running a ubuntu:20.04 container on an older version of ubuntu (16.04), which has a 4.x kernel, so the issue in that case was that the container is making syscalls that the host didn't understand (or at least libseccomp on the host)
>
> https://github.com/moby/moby/issues/40734#issuecomment-638267023

Also, https://github.com/nodejs/docker-node/issues/1780#issuecomment-1267370041

> TL;DR: you'll need to update your host to:
> - Docker version 19.03.9 or newer
> - libseccomp version 2.4.2 or newer


Just for future users.
